plutus-core/plutus-core/src/PlutusCore/Subst.hs

```diff
@@ -147,65 +147,46 @@ termSubstFreeNames = purely termSubstFreeNamesA
-fvTerm :: Ord name => Term tyname name uni fun ann -> Set name
-fvTerm = cata f
-  where
-    f (VarF _ n)        = singleton n
-    f (TyAbsF _ _ _ t)  = t
-    f (LamAbsF _ n _ t) = delete n t
-    f (ApplyF _ t1 t2)  = t1 `union` t2
-    f (TyInstF _ t _)   = t
-    f (UnwrapF _ t)     = t
-    f (IWrapF _ _ _ t)  = t
-    f ConstantF{}       = Set.empty
-    f BuiltinF{}        = Set.empty
-    f ErrorF{}          = Set.empty
+fvTerm :: Ord name => Traversal' (Term tyname name uni fun ann) name
+fvTerm = fvTermCtx mempty
+
+fvTermCtx :: Ord name => Set.Set name -> Traversal' (Term tyname name uni fun ann) name
+fvTermCtx bound f = \case
+    Var a n         -> Var a <$> (if Set.member n bound then pure n else f n)
+    LamAbs a n ty t -> LamAbs a n ty <$> fvTermCtx (Set.insert n bound) f t
+    t               -> (termSubterms . fvTermCtx bound) f t
 
 -- | Get all the free type variables in a term.
-ftvTerm :: Ord tyname => Term tyname name uni fun ann -> Set tyname
-ftvTerm = cata f
-  where
-    f (TyAbsF _ ty _ t)    = delete ty t
-    f (LamAbsF _ _ ty t)   = ftvTy ty `union` t
-    f (ApplyF _ t1 t2)     = t1 `union` t2
-    f (TyInstF _ t ty)     = t `union` ftvTy ty
-    f (UnwrapF _ t)        = t
-    f (IWrapF _ pat arg t) = ftvTy pat `union` ftvTy arg `union` t
-    f (ErrorF _ ty)        = ftvTy ty
-    f VarF{}               = Set.empty
-    f ConstantF{}          = Set.empty
-    f BuiltinF{}           = Set.empty
+ftvTerm :: Ord tyname => Traversal' (Term tyname name uni fun ann) tyname
+ftvTerm = ftvTermCtx mempty
+
+ftvTermCtx :: Ord tyname => Set.Set tyname -> Traversal' (Term tyname name uni fun ann) tyname
+ftvTermCtx bound f = \case
+    TyAbs a ty k t -> TyAbs a ty k <$> ftvTermCtx (Set.insert ty bound) f t
+    -- sound because the subterms and subtypes are disjoint
+    t              -> ((termSubterms . ftvTermCtx bound) `Unsound.adjoin` (termSubtypes . ftvTyCtx bound)) f t
 
 -- | Get all the free type variables in a type.
-ftvTy :: Ord tyname => Type tyname uni ann -> Set tyname
-ftvTy = cata f
-  where
-    f (TyVarF _ ty)          = singleton ty
-    f (TyFunF _ i o)         = i `union` o
-    f (TyIFixF _ pat arg)    = pat `union` arg
-    f (TyForallF _ bnd _ ty) = delete bnd ty
-    f (TyLamF _ bnd _ ty)    = delete bnd ty
-    f (TyAppF _ ty1 ty2)     = ty1 `union` ty2
-    f TyBuiltinF{}           = Set.empty
+ftvTy :: Ord tyname => Traversal' (Type tyname uni ann) tyname
+ftvTy = ftvTyCtx mempty
+
+ftvTyCtx :: Ord tyname => Set.Set tyname -> Traversal' (Type tyname uni ann) tyname
+ftvTyCtx bound f = \case
+    TyVar a ty          -> TyVar a <$> (if Set.member ty bound then pure ty else f ty)
+    TyForall a bnd k ty -> TyForall a bnd k <$> ftvTyCtx (Set.insert bnd bound) f ty
+    TyLam a bnd k ty    -> TyLam a bnd k <$> ftvTyCtx (Set.insert bnd bound) f ty
+    t                   -> (typeSubtypes . ftvTyCtx bound) f t
+
 
+-- TODO: these could be Traversals
```

Comment:
what do you mean by this TODO? do you want to make them traversals but it is diffcult or you are still contemplating restricting them to be only folds?